### DISCUSSION TITLE
Nested constants don't seem to be handled properly

### DISCUSSION THREAD

**Discussion by findhumane** [lsegal/yard#1558] [Ideas]
[One common technique](https://stackoverflow.com/a/5675566/4135310) to define enumerations in Ruby that I like does not seem to be handled by YARD properly.

## Steps to reproduce

This is the minimal reproduction for the issue. I've done my best to remove
all extraneous code and unique environment state on my machine before providing
these steps:

1. Create `yard_nested_constants.rb`:
   ```ruby
   class YardNestedConstants
     # Attributes for products
     PRODUCT_ATTRIBUTES = [
   
       # The name of the product.
       PRODUCT_NAME = :product_name,
   
       # The name of the producer.
       PRODUCER = :producer,
     ]
   end
   ```
1. Run the following command:
   ```
   yard doc
   ```
1. Open browser to `doc/YardNestedConstants.html`

## Actual Output

![Screenshot_20230828_225721](https://github.com/lsegal/yard/assets/111241339/70ebd5b8-14d7-409f-9a51-753c49b54c36)

## Expected Output

Documentation of `PRODUCT_NAME` and `PRODUCER` should be processed.

## Environment details:

- OS: Fedora Linux 37
- Ruby version (`ruby -v`): `ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [x86_64-linux]`
- YARD version (`yard -v`): `yard 0.9.34`

I have read the [Contributing Guide][contrib].

[contrib]: https://github.com/lsegal/yard/blob/main/CONTRIBUTING.md

**Comment by lsegal** [comment]
By default YARD only processes top-level statements for performance as well as correctness reasons. As a workaround, it wouldn't be too difficult to write a plugin to automatically handle this case:

```ruby
# place in a file like `.yard/ext/array_constant_handler.rb`
class ArrayConstantHandler < YARD::Handlers::Ruby::Base
  handles :assign
  namespace_only

  process do
    arr = statement[1]
    arr.children.each {|c| parse_block(c) } if arr.type == :array
  end
end
```

Place `-e .yard/ext/array_constant_handler.rb` in your `.yardopts` or pass that arg to `yard doc` on the command line.

You can package that up in a gem that has the `yard-YOURNAME` prefix and can then use `--plugin YOURNAME` instead of the -e argument (assuming the gem is installed).

You could also build this out as a [Directive](https://rubydoc.info/gems/yard/0.9.27/YARD/Tags/Directive) to be a bit more surgical about when this logic gets invoked. It would be similarly simple.

It's currently unclear to me if this is something that is widely needed in YARD core; this is the first time it's popped up as a feature request. Experimenting with a plugin could identify how commonly used the idiom is in the wild, and if it is, it could be merged in.

**Comment by findhumane** [comment]
That worked, thanks!
